### PR TITLE
List endpoint

### PR DESCRIPTION
**Pull request by sixlive**
## Description

❗This PR is built off of #10 

- Adds the list endpoint functionality.
- Breaking: Update all methods to respond in an associative array rather than a `json_decode` object
  - Adjust the geocoding tests accordingly
  - Groups the geocoding tests by forward/reverse using the `describe` method

### CODE REVIEW

File: src/Geocodio.php [hunk 1]
```diff
@@ -37,7 +39,25 @@ class Geocodio
...
+        $this->apiKey = getenv('GEOCODIO_API_KEY');
+
+        if ($hostname = getenv('GEOCODIO_HOSTNAME')) {
+            $this->hostname = $hostname;
+        }
+
+        if ($apiVersion = getenv('GEOCODIO_API_VERSION')) {
+            $this->apiVersion = $apiVersion;
+        }
+    }
```
Comment (sixlive) [inline src/Geocodio.php:60]: Not sold on this but I kind of like the idea of defaulting to the ENV vars. I think I might refactor this to constructor ags (updating the service provider) and setting up client overrides via `->withClient(Client $client)`

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,11 @@
 
 All notable changes to `geocodio-library-php` will be documented in this file
 
+## 2.0 - TBD
+
+- Added support for list geocoding
+- Breaking: all method responses now return an associative array rather than a JSON decoded object
```
Comment (MiniCodeMonkey) [inline CHANGELOG.md:8]: Can also mention here that minimum PHP version is now 8.2

File: bin/git-hooks/formatting [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    _print "Running rector..."
+    vendor/bin/rector process --no-diffs
+
+    files_after_format=$(git diff --name-only --diff-filter=d)
+
+    # Find files fixed by pint by comparing file lists before and after pint run
+    files_fixed_by_format=$(comm -13 <(sort <<<"$files_before_format") <(sort <<<"$files_after_format"))
+
+    # Re-stage files fixed by pint
```
Comment (MiniCodeMonkey) [inline bin/git-hooks/formatting:21]: Nice!!

File: composer.json [hunk 4]
```diff
@@ -25,7 +25,8 @@
         "phpunit/phpunit": "^11.0",
         "projektgopher/whisky": "^0.7.0",
         "rector/rector": "^1.2",
-        "symfony/var-dumper": "^4.3"
+        "symfony/var-dumper": "^4.3",
+        "spatie/ray": "^1.41"
```
Comment (MiniCodeMonkey) [inline composer.json]: 👍

File: ray.php [hunk 5]
```diff
@@ -0,0 +1,32 @@
+<?php
```
Comment (MiniCodeMonkey) [inline ray.php]: Is this `ray` config file required. Or can we omit it to use default settings (just using environment variables to overwrite)

File: src/Concerns/SendsRequests.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            $this->getHeaders(),
+            $options[RequestOptions::HEADERS] ?? [],
+        );
+
+        return $options;
+    }
+
+    protected function formatUrl(string $endpoint): string
+    {
+        return vsprintf('https://%s/%s/%s', [
```
Comment (MiniCodeMonkey) [inline src/Concerns/SendsRequests.php:69]: TIL: There's a `vsprintf(...)` :)

File: src/Concerns/SendsRequests.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            }
+        }
+
+        return $query;
+    }
+
+    protected function getHeaders(): array
+    {
+        return [
+            'User-Agent' => sprintf('geocodio-library-php/%s', Geocodio::SDK_VERSION),
```
Comment (MiniCodeMonkey) [inline src/Concerns/SendsRequests.php:92]: 🙌

File: src/Geocodio.php [hunk 8]
```diff
@@ -37,7 +41,23 @@ class Geocodio
         'country',
     ];
 
-    public function __construct(private readonly Client $client = new Client) {}
+    /**
+     * @var Current SDK version
+     */
+    const SDK_VERSION = '1.2.0';
```
Comment (MiniCodeMonkey) [inline src/Geocodio.php]: Can bump this to 2.0.0 before merging in

File: src/Geocodio.php [hunk 9]
```diff
@@ -60,138 +80,253 @@ public function setApiVersion(string $apiVersion): self
...
-            }
+        return true;
+    }
 
-            $errorCode = $response->getStatusCode();
+    protected function uploadMultipartFile(
+        string $fileContents,
+        GeocodeDirection $direction,
+        string $format,
+        string $callbackWebhook,
```
Comment (MiniCodeMonkey) [inline src/Geocodio.php]: Should this be `?string $callbackWebhook = null` instead. Since it's optional?

File: src/Geocodio.php [hunk 10]
```diff
@@ -60,138 +80,253 @@ public function setApiVersion(string $apiVersion): self
...
+        ?string $filename = null
+    ): Response {
+        if (is_file($fileContents) && ! file_exists($fileContents)) {
+            throw GeocodioException::fileNotFound($fileContents);
```
Comment (MiniCodeMonkey) [inline src/Geocodio.php:302]: Neat!!

**Review by MiniCodeMonkey — APPROVED** [review]

File: ray.php (hunk 5 above)
Comment (sixlive) [reply ray.php]: I'm not 100% sure but I think the config file is required. I actually didn't intend on shipping with ray as part of the package so I'm going to update to remove it.

File: src/Concerns/SendsRequests.php (hunk 6 above)
Comment (sixlive) [reply src/Concerns/SendsRequests.php:69]: I love `sprintf` and `vsprintf` I use them all the time for string formatting.

File: src/Geocodio.php (hunk 8 above)
Comment (sixlive) [reply src/Geocodio.php]: Yeah! I was going to update it before tagging but I'll do it here!

File: src/Geocodio.php (hunk 9 above)
Comment (sixlive) [reply src/Geocodio.php]: Good call!

**Review by MiniCodeMonkey — APPROVED** [review]

**Pull request merged**